### ISSUE TITLE
Tensor roll with different shifts

### ISSUE THREAD

**Issue by jimouris** [pytorch/pytorch#125422]
### 🚀 The feature, motivation and pitch

Hello 👋

It seems that `roll` only takes integers or tuples and applies the same roll in each element of the tensor. Is it possible to extend `roll` to take a tensor as an argument and shift each position of the tensor a different amount?

Something like:
```python
import torch

size = 10

shifts = torch.tensor([1, 2, -1])

one_hot_indices = torch.tensor([1, 3, 4])
one_hot = torch.nn.functional.one_hot(one_hot_indices, size)
print(one_hot)

# one_hot.roll(shifts)
```

cc: @nilmemo

**Comment by janeyx99** [comment]
Could you elaborate on your use case to motivate this request?

**Comment by jimouris** [comment]
Sure @janeyx99! I skipped it initially since it's a bit convoluted but here it goes :) 

I'm working on adding a lookup table functionality on [CrypTen](https://github.com/facebookresearch/CrypTen), a framework built on top of pytorch for privacy-preserving machine learning. Being able to `roll` with different shifts is crucial for performance (I currently have to do a sequential loop).

The idea is the following:
* you have an encrypted value `x`,
* `x` gets masked with a random value `r` as `shift = x + r` and gets revealed,
* `r` is also converted to a 1-hot vector `r_vec`.
* we can now use the `shift` to roll `r_vec` and then use this in a dot product with a public lookup table `T`. 
* The result is the value at the `x`th position of `T` while we don't learn `x`.
* `x` can now be a tensor, and we want different `r` values for each index, hence the different shifts.

Hope this helps!

**Comment by Manvelloid** [comment]
This feature would be very useful

**Comment by manel1874** [comment]
Indeed @Manvelloid. @janeyx99 any idea when this will be supported? Thank you!

**Comment by nilmemo** [comment]
+1

**Comment by albanD** [comment]
Is this something being considered by the array API @rgommers https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.roll.html ?

Needs design to see if we want this to be a different op or not?
And the exact API: are we just extending from list(int) to tensor of integer dtype?

**Comment by rgommers** [comment]
It isn't. I haven't seen this request for different shifts along the same dimension before I think; I searched the NumPy issue tracker as well and didn't find anything there.

> Needs design to see if we want this to be a different op or not?

Since it's already possible for `roll` to accept a tuple of ints with a shift per `dims` (e.g., `torch.roll(one_hot, (1, 3), dims=(0, 1))`), it seems potentially confusing if a tensor of ints started to be accepted for `shifts` as well but had a different meaning than a tuple of ints. So this feels more like a request for `roll_along_dim`.
